lib/ansible/galaxy/collection/__init__.py

```diff
@@ -233,7 +241,7 @@ def build_collection(collection_path, output_path, force):
         ver=collection_meta['version'],
     )
     b_collection_output = os.path.join(
-        output_path,
+        to_bytes(output_path, errors='surrogate_or_strict'),
```

Comment:
@s-hertel how did you hit this? I've set `bytes` in typing annotations but clearly, that was just an assumption based on what I tracked by just looking at code. If we can figure out what causes this to be unicode, we'll be able to handle this better.

Reply:
I hit it with the first task in [test/integration/targets/ansible-galaxy-collection/tasks/build.yml](https://github.com/ansible/ansible/blob/devel/test/integration/targets/ansible-galaxy-collection/tasks/build.yml#L2). We have the bytes value, but don't use it when calling build_collection https://github.com/ansible/ansible/blob/devel/lib/ansible/cli/galaxy.py#L792.

Traceback if I comment out this line:
```
fatal: [testhost]: FAILED! => {
    "changed": true,
    "cmd": [
        "ansible-galaxy",
        "collection",
        "build",
        "-vvvv"
    ],
    "delta": "0:00:00.483686",
    "end": "2020-11-19 17:46:17.230153",
    "invocation": {
        "module_args": {
            "_raw_params": "ansible-galaxy collection build -vvvv",
            "_uses_shell": false,
            "argv": null,
            "chdir": "/tmp/ansible.jl8257mv.test/galaxy/scratch/ansible_test/my_collection",
            "creates": null,
            "executable": null,
            "removes": null,
            "stdin": null,
            "stdin_add_newline": true,
            "strip_empty_ends": true,
            "warn": false
        }
    },
    "msg": "non-zero return code",
    "rc": 250,
    "start": "2020-11-19 17:46:16.746467",
    "stderr": "\u001b[0;31mERROR! Unexpected Exception, this is probably a bug: Can't mix strings and bytes in path components\u001b[0m",
    "stderr_lines": [
        "\u001b[0;31mERROR! Unexpected Exception, this is probably a bug: Can't mix strings and bytes in path components\u001b[0m"
    ],
    "stdout": "\u001b[0;34mansible-galaxy 2.11.0.dev0\u001b[0m\n\u001b[0;34m  config file = /root/ansible/test/results/.tmp/integration/ansible-galaxy-collection-qi4a7zn7-ÅÑŚÌβŁÈ/test/integration/integration.cfg\u001b[0m\n\u001b[0;34m  configured module search path = ['/root/ansible/test/support/integration/plugins/modules']\u001b[0m\n\u001b[0;34m  ansible python module location = /root/ansible/lib/ansible\u001b[0m\n\u001b[0;34m  ansible collection location = /root/.ansible/collections:/usr/share/ansible/collections\u001b[0m\n\u001b[0;34m  executable location = /root/ansible/bin/ansible-galaxy\u001b[0m\n\u001b[0;34m  python version = 3.6.9 (default, Jul 17 2020, 12:50:27) [GCC 8.4.0]\u001b[0m\n\u001b[0;34m  jinja version = 2.11.2\u001b[0m\n\u001b[0;34m  libyaml = True\u001b[0m\n\u001b[0;34mUsing /root/ansible/test/results/.tmp/integration/ansible-galaxy-collection-qi4a7zn7-ÅÑŚÌβŁÈ/test/integration/integration.cfg as config file\u001b[0m\n\u001b[0;34mSkipping '/tmp/ansible.jl8257mv.test/galaxy/scratch/ansible_test/my_collection/galaxy.yml' for collection build\u001b[0m\nthe full traceback was:\n\nTraceback (most recent call last):\n  File \"/root/ansible/bin/ansible-galaxy\", line 135, in <module>\n    exit_code = cli.run()\n  File \"/root/ansible/lib/ansible/cli/galaxy.py\", line 523, in run\n    context.CLIARGS['func']()\n  File \"/root/ansible/lib/ansible/cli/galaxy.py\", line 801, in execute_build\n    build_collection(collection_path, output_path, force)\n  File \"/root/ansible/lib/ansible/galaxy/collection/__init__.py\", line 246, in build_collection\n    to_bytes(artifact_tarball_file_name, errors='surrogate_or_strict'),\n  File \"/usr/lib/python3.6/posixpath.py\", line 94, in join\n    genericpath._check_arg_types('join', a, *p)\n  File \"/usr/lib/python3.6/genericpath.py\", line 151, in _check_arg_types\n    raise TypeError(\"Can't mix strings and bytes in path components\") from None\nTypeError: Can't mix strings and bytes in path components",
    "stdout_lines": [
        "\u001b[0;34mansible-galaxy 2.11.0.dev0\u001b[0m",
        "\u001b[0;34m  config file = /root/ansible/test/results/.tmp/integration/ansible-galaxy-collection-qi4a7zn7-ÅÑŚÌβŁÈ/test/integration/integration.cfg\u001b[0m",
        "\u001b[0;34m  configured module search path = ['/root/ansible/test/support/integration/plugins/modules']\u001b[0m",
        "\u001b[0;34m  ansible python module location = /root/ansible/lib/ansible\u001b[0m",
        "\u001b[0;34m  ansible collection location = /root/.ansible/collections:/usr/share/ansible/collections\u001b[0m",
        "\u001b[0;34m  executable location = /root/ansible/bin/ansible-galaxy\u001b[0m",
        "\u001b[0;34m  python version = 3.6.9 (default, Jul 17 2020, 12:50:27) [GCC 8.4.0]\u001b[0m",
        "\u001b[0;34m  jinja version = 2.11.2\u001b[0m",
        "\u001b[0;34m  libyaml = True\u001b[0m",
        "\u001b[0;34mUsing /root/ansible/test/results/.tmp/integration/ansible-galaxy-collection-qi4a7zn7-ÅÑŚÌβŁÈ/test/integration/integration.cfg as config file\u001b[0m",
        "\u001b[0;34mSkipping '/tmp/ansible.jl8257mv.test/galaxy/scratch/ansible_test/my_collection/galaxy.yml' for collection build\u001b[0m",
        "the full traceback was:",
        "",
        "Traceback (most recent call last):",
        "  File \"/root/ansible/bin/ansible-galaxy\", line 135, in <module>",
        "    exit_code = cli.run()",
        "  File \"/root/ansible/lib/ansible/cli/galaxy.py\", line 523, in run",
        "    context.CLIARGS['func']()",
        "  File \"/root/ansible/lib/ansible/cli/galaxy.py\", line 801, in execute_build",
        "    build_collection(collection_path, output_path, force)",
        "  File \"/root/ansible/lib/ansible/galaxy/collection/__init__.py\", line 246, in build_collection",
        "    to_bytes(artifact_tarball_file_name, errors='surrogate_or_strict'),",
        "  File \"/usr/lib/python3.6/posixpath.py\", line 94, in join",
        "    genericpath._check_arg_types('join', a, *p)",
        "  File \"/usr/lib/python3.6/genericpath.py\", line 151, in _check_arg_types",
        "    raise TypeError(\"Can't mix strings and bytes in path components\") from None",
        "TypeError: Can't mix strings and bytes in path components"
    ]
}
```

I like the `b_` convention for variables that are bytes, makes it easy to scan the code.

Reply:
Agreed. `b_` is important but this convention isn't always being followed and there's no automatic way to verify what's expected. Explicit type annotations help with this but only if they are set correctly and I've either made a wrong assumption or this function is called with both bytes and unicode from different places. That's what needs investigation.

Reply:
Ah. It also used to be called from download_collections as unicode also. But in your PR you're passing in bytes instead: https://github.com/ansible/ansible/pull/72591/files#diff-b1b46df93a963b73e0c4a4c7a598a35d3135076a5ecf23f1c07323be0005a5fcR313
```
            if concrete_coll_pin.is_dir:
                b_dest_path = to_bytes(
                    build_collection(
                        b_src_path, b_output_path, force=True,
                    ),
                    errors='surrogate_or_strict',
                )
```

Reply:
yep. and it's better to use one type everywhere. it seems like it's better to always use bytes for paths so I tried to do so.

Reply:
Yep, you changed it to multiple types. I don't care what type we use, but if you're changing it to bytes please add b_ for clarity.